### PR TITLE
useContext

### PR DESCRIPTION
**Pull request by marcosrdz**

### CODE REVIEW

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/wza3y8f43xdw9b9mgw0kbhywrm

[download apk](http://filestorage.bluewallet.io:1488/1602689275.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/hu79pvdjj6zm1wyh56t2d2cmcw

[download apk](http://filestorage.bluewallet.io:1488/1602718791.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/8xmnevvhr4cpxz2tcxcyhg7knr

[download apk](http://filestorage.bluewallet.io:1488/1602723103.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/4x6ery8k19kyjdnpkv56rhddcc

[download apk](http://filestorage.bluewallet.io:1488/1602726439.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/257y8gtp2q0e8z7f9ncw2pxhv8

[download apk](http://filestorage.bluewallet.io:1488/1602727626.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/kpb16eedee6xdfvjg2rxr25tj8

[download apk](http://filestorage.bluewallet.io:1488/1602731509.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/p9zvkupf86agcjgbpv8m59buqm

[download apk](http://filestorage.bluewallet.io:1488/1602734482.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/zf0z43v8a4a2pn09z9z4m0b8zg

[download apk](http://filestorage.bluewallet.io:1488/1602764442.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/162azfw1gq6gkapfgr63x8gm3r

[download apk](http://filestorage.bluewallet.io:1488/1602770615.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/ux886qmymzu3f69hzdrbx87tt0

[download apk](http://filestorage.bluewallet.io:1488/1602776792.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/qmrz7kcxf150fp5hznc3ewr9jc

[download apk](http://filestorage.bluewallet.io:1488/1602781910.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/1mfvbhy7u0xhucubgvfy6mzrhg

[download apk](http://filestorage.bluewallet.io:1488/1602814310.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/ph1ze4f2drw7kcreg3jcbhr0u4

[download apk](http://filestorage.bluewallet.io:1488/1602815515.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/bpank73d9gavu93f6fkgkw88u0

[download apk](http://filestorage.bluewallet.io:1488/1602822859.apk)

**Review by limpbrains — DISMISSED** [review]
Great work! This approach of using context to get fresh data from wallets feels more react than weird events system. 

Do you think there could be performance issues, since now you can have a few components subscribed to all changes?

The way a few classes like `DeeplinkSchemaMatch` or `WalletImport` where refactored feels odd. Because you add methods to react components and put business logic inside them. 
Idea to bound everything using react API feels weird to me. It is possible to archive something like this using EventEmitter somewhere inside `BlueApp`. I prefer keep react UI and business logic separated.

**Comment by marcosrdz** [comment]
> Great work! This approach of using context to get fresh data from wallets feels more react than weird events system. 
> 
> 
> 
> Do you think there could be performance issues, since now you can have a few components subscribed to all changes?
> 
> 
> 
> The way a few classes like `DeeplinkSchemaMatch` or `WalletImport` where refactored feels odd. Because you add methods to react components and put business logic inside them. 
> 
> Idea to bound everything using react API feels weird to me. It is possible to archive something like this using EventEmitter somewhere inside `BlueApp`. I prefer keep react UI and business logic separated. 

Xcode is reporting a lower consumption of memory. And the app behaves faster. The "business logic" will be moved to useReducer later.
I also have to place those items in the "DOM" so they have access to the context and are able to answer to state changes instantly anywhere(look at watchconnectivity for example). With this change, i can work with my wallets and every change gets synced to my watch through hooks..and this is all without blueapp being aware of that component.

File: class/wallet-import.js [hunk 1]
```diff
@@ -50,39 +48,48 @@ export default class WalletImport {
...
-    const placeholderWalletIndex = BlueApp.wallets.findIndex(wallet => wallet.type === PlaceholderWallet.type);
-    if (placeholderWalletIndex > -1) {
-      BlueApp.wallets.splice(placeholderWalletIndex, 1);
+  WalletImport.isWalletImported = secret => {
+    const wallet = wallets.some(wallet => wallet.getSecret() === secret && wallet.type !== PlaceholderWallet.type);
+    if (wallet) {
+      return true;
     }
-  }
+    return false;
```
Comment (Overtorment) [inline class/wallet-import.js:67]: can be reduced to  `return !!wallet`

File: class/wallet-import.js [hunk 2]
```diff
@@ -17,30 +16,29 @@ import {
...
    * @private
    */
-  static async _saveWallet(w, additionalProperties) {
-    const wallet = BlueApp.getWallets().some(
-      wallet => (wallet.getSecret() === w.secret || wallet.getID() === w.getID()) && wallet.type !== PlaceholderWallet.type,
-    );
-    if (wallet) {
-      alert('This wallet has been previously imported.');
+  WalletImport._saveWallet = async (w, additionalProperties) => {
+    if (WalletImport.isWalletImported(w.getSecret())) {
```
Comment (Overtorment) [inline class/wallet-import.js:38]: comparing secrets is not enough, you need to compare IDs as well
(for example, multisig wallet takes provided secret and transfors it in internal representation. and  `getSectret` outputs it in different format, so technically different secrets could still create same wallets with proper equal IDs)

File: screen/settings/GeneralSettings.js [hunk 3]
```diff
@@ -14,14 +13,15 @@ const styles = StyleSheet.create({
 });
 
 const GeneralSettings = () => {
+  const { isAdancedModeEnabled, setIsAdancedModeEnabled, wallets } = useContext(BlueStorageContext);
   const [isLoading, setIsLoading] = useState(true);
-  const [isAdancedModeEnabled, setIsAdancedModeEnabled] = useState(false);
+  const [adancedIsModeEnabled, setAdancedModeIsEnabled] = useState(false);
```
Comment (Overtorment) [inline screen/settings/GeneralSettings.js:18]: mindfuck with names. can it be improved?

File: tests/integration/notifications.test.js [hunk 4]
```diff
@@ -1,11 +1,11 @@
 /* global it, describe */
 const assert = require('assert');
-const notifications = require('../../blue_modules/notifications');
+const { default: BlueNotifications } = require('../../blue_modules/notifications');
```
Comment (Overtorment) [inline tests/integration/notifications.test.js:3]: unnecessary rename

**Review by Overtorment — CHANGES_REQUESTED** [review]
my main concern is whethere my IDE gona understand code correctly and provide code navigation & code completion.

for example, that `this.context` provides a bunch of functions from BlueApp


or that in  `const { txMetadata, wallets } = useContext(BlueStorageContext);` 
`wallets` is an array of AbstractWallet


I havent tested it with my IDE yet, but in case it wont recognize it. is there a way to make a typehints for ide? 
for example, you might notice that i have things like
`const BlueApp: AppStorage = require('../../BlueApp');` 
or 
```
    /** @type {AbstractHDElectrumWallet|WatchOnlyWallet} */
    const wallet = this.state.fromWallet;
```


PS. e2e fails

File: BlueComponents.js [hunk 5]
```diff
@@ -275,7 +277,7 @@ export class BlueWalletNavigationHeader extends Component {
 
     wallet.hideBalance = !wallet.hideBalance;
     this.setState({ wallet });
-    await BlueApp.saveToDisk();
+    await this.context.saveToDisk();
```
[the remark this answers is not in the document]
Comment (marcosrdz) [reply BlueComponents.js:280]: this.context is defined in React.Component. Just like a setState and props

File: blue_modules/BlueStorage.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+export const BlueStorageProvider = ({ children }) => {
+  const [wallets, setWallets] = useState([]);
+  const [pendingWallets, setPendingWallets] = useState([]);
+  const [selectedWallet, setSelectedWallet] = useState('');
+  const [walletsInitialized, setWalletsInitialized] = useState(false);
+  const saveToDisk = async () => {
+    BlueApp.tx_metadata = txMetadata;
+    await BlueApp.saveToDisk();
+    setWallets([...BlueApp.getWallets()]);
+    txMetadata = BlueApp.tx_metadata;
```
[the remark this answers is not in the document]
Comment (marcosrdz) [reply blue_modules/BlueStorage.js:16]: Lint was complaining it tx_metadata

File: blue_modules/BlueStorage.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  const createFakeStorage = BlueApp.createFakeStorage;
+  const decryptStorage = BlueApp.decryptStorage;
+  const isDeleteWalletAfterUninstallEnabled = BlueApp.isDeleteWalletAfterUninstallEnabled;
+  const setResetOnAppUninstallTo = BlueApp.setResetOnAppUninstallTo;
+  const isPasswordInUse = BlueApp.isPasswordInUse;
+  const cachedPassword = BlueApp.cachedPassword;
+  const setIsAdancedModeEnabled = BlueApp.setIsAdancedModeEnabled;
+  const getHodlHodlSignatureKey = BlueApp.getHodlHodlSignatureKey;
+  const addHodlHodlContract = BlueApp.addHodlHodlContract;
+  const getHodlHodlContracts = BlueApp.getHodlHodlContracts;
```
[the remark this answers is not in the document]
Comment (marcosrdz) [reply blue_modules/BlueStorage.js:80]: Thats correct. Any function or property that needs exposure should be defined here

File: blue_modules/events.js [hunk 8]
```diff
@@ -1,47 +0,0 @@
-function EV(eventName, arg, isExclusive) {
```
[the remark this answers is not in the document]
Comment (marcosrdz) [reply blue_modules/events.js:1]: The end of an era

File: App.js [hunk 9]
```diff
@@ -27,20 +27,22 @@ import loc from './loc';
 import { BlueDefaultTheme, BlueDarkTheme, BlueCurrentTheme } from './components/themes';
 import InitRoot from './Navigation';
 import BlueClipboard from './blue_modules/clipboard';
+import { BlueStorageContext } from './blue_modules/BlueStorage';
+import WatchConnectivity from './WatchConnectivity';
+import DeviceQuickActions from './class/quick-actions';
+import BlueNotifications from './blue_modules/notifications';
```
[the remark this answers is not in the document]
Comment (marcosrdz) [reply App.js]: i placed the notification code inside a function so i can pass a props to it..i renamed it back to notifications.

File: screen/settings/GeneralSettings.js (hunk 3 above)
Comment (marcosrdz) [reply screen/settings/GeneralSettings.js]: changed to isAdancedModeSwitchEnabled

**Comment by marcosrdz** [comment]
>    /** @type {AbstractHDElectrumWallet|WatchOnlyWallet} */
>     const wallet = this.state.fromWallet;



> my main concern is whethere my IDE gona understand code correctly and provide code navigation & code completion.
> 
> for example, that `this.context` provides a bunch of functions from BlueApp
> 
> or that in `const { txMetadata, wallets } = useContext(BlueStorageContext);`
> `wallets` is an array of AbstractWallet
> 
> I havent tested it with my IDE yet, but in case it wont recognize it. is there a way to make a typehints for ide?
> for example, you might notice that i have things like
> `const BlueApp: AppStorage = require('../../BlueApp');`
> or
> 
> ```
>     /** @type {AbstractHDElectrumWallet|WatchOnlyWallet} */
>     const wallet = this.state.fromWallet;
> ```
> 
> PS. e2e fails

Those never had autocomplete for me. Your IDE recognizes those but VS Code doesnt. I'll have to check.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/z7gpnvez4ef18j20dr36na3vwm

[download apk](http://filestorage.bluewallet.io:1488/1602917265.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/9d5k0m4v1nxck96tm3kh8aj46m

[download apk](http://filestorage.bluewallet.io:1488/1602942586.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/adjtw04zhquaf5xvtmy0tfrxjg

[download apk](http://filestorage.bluewallet.io:1488/1602947106.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/72y4b4x4wxvjzeuj6haf7eem2r

[download apk](http://filestorage.bluewallet.io:1488/1602953169.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/5h2zbve416z3vdp8zthny0xgm4

[download apk](http://filestorage.bluewallet.io:1488/1602993369.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/kvgw9x6qm1mrdffmwjur8xbmv8

[download apk](http://filestorage.bluewallet.io:1488/1603073624.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/wv4gec0czc80a6tyz7k827qtf8

[download apk](http://filestorage.bluewallet.io:1488/1603077303.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/b1mcayu1x8kjpbt8jdr3fg4d84

[download apk](http://filestorage.bluewallet.io:1488/1603078133.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/e8kypefadbe9tx0rcmjw020xtm

[download apk](http://filestorage.bluewallet.io:1488/1603113186.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/v50fgqx18v4n2kt2w7arp7ema0

[download apk](http://filestorage.bluewallet.io:1488/1603118041.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/5rz79ucb7zdgh0m2ffatuuqtq4

[download apk](http://filestorage.bluewallet.io:1488/1603159489.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/cuxukh0nt030fj0tn1pqmcz4m8

[download apk](http://filestorage.bluewallet.io:1488/1603164998.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/09wn52aba4vh8a1gtguunuh3dg

[download apk](http://filestorage.bluewallet.io:1488/1603167932.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/mqx4rpcz1ncq8yt5fr4mb36hbg

[download apk](http://filestorage.bluewallet.io:1488/1603207465.apk)

**Review by Overtorment — DISMISSED** [review]
ok if @ncoelho has no objections and cant find any bugs we can merge it

**Comment by ncoelho** [comment]
Weird behaviour

**Comment by ncoelho** [comment]
Weird first load

**Comment by ncoelho** [comment]
Wallet creation doesn't go to latest wallet

**Comment by ncoelho** [comment]
transactions.js is not triggering a refresh after doing a transaction. 

☝️ this is a blocker...

**Comment by marcosrdz** [comment]
> Wallet creation doesn't go to latest wallet
> 
> 

Fixed

**Comment by marcosrdz** [comment]
> Weird behaviour
> 
> 

Fixed

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/hq2nzzf87h3jrtu1j9hwu6h85g

[download apk](http://filestorage.bluewallet.io:1488/1603290827.apk)

**Comment by ncoelho** [comment]
> > Wallet creation doesn't go to latest wallet
> > 
> 
> Fixed

tACK. This looks good

**Comment by ncoelho** [comment]
> > Weird behaviour
> > 
> 
> Fixed

I still see the bad loading on the listing, at least on onchain wallets.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/n526y0cym7nu74j5e3j5uttv5g

[download apk](http://filestorage.bluewallet.io:1488/1603321859.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/yb3ve0tqhxec6jwj4tet3mx7hg

[download apk](http://filestorage.bluewallet.io:1488/1603412031.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/6n7cwu81hcdfxr59qbrn011q0r

[download apk](http://filestorage.bluewallet.io:1488/1603415909.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/yw7w57tj5fp49e8a0umnq25020

[download apk](http://filestorage.bluewallet.io:1488/1603483314.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/e49gjyv0d58xhafqh1tc07502g

[download apk](http://filestorage.bluewallet.io:1488/1603490524.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/k44gwr04dt8tza31dzwdhzy8br

[download apk](http://filestorage.bluewallet.io:1488/1603495128.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/qepqt5pg3emubq2twz1c3ypkm4

[download apk](http://filestorage.bluewallet.io:1488/1603501001.apk)

**Pull request merged**